frontend/gateway/gateway.go

```diff
@@ -738,6 +747,406 @@ func (lbf *llbBridgeForwarder) Inputs(ctx context.Context, in *pb.InputsRequest)
 	}, nil
 }
 
+func (lbf *llbBridgeForwarder) NewContainer(ctx context.Context, in *pb.NewContainerRequest) (_ *pb.NewContainerResponse, err error) {
+	logrus.Debugf("|<--- NewContainer %s", in.ContainerID)
+	ctrReq := NewContainerRequest{
+		ContainerID:  in.ContainerID,
+		NetMode:      in.Network,
+		SecurityMode: in.Security,
+	}
+
+	for _, m := range in.Mounts {
+		refProxy, err := lbf.convertRef(m.ResultID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to find ref %s for %q mount", m.ResultID, m.Dest)
+		}
+		ctrReq.Mounts = append(ctrReq.Mounts, Mount{
+			Dest:      m.Dest,
+			Selector:  m.Selector,
+			Readonly:  m.Readonly,
+			MountType: m.MountType,
+			RefProxy:  refProxy,
+		})
+	}
+
+	// Not using `ctx` here because it will get cancelled as soon as NewContainer returns
+	// and we want the context to live for the duration of the container.
+	ctr, err := NewContainer(context.Background(), lbf.llbBridge, ctrReq)
+	if err != nil {
+		return nil, stack.Enable(err)
+	}
+
+	lbf.ctrsMu.Lock()
+	// ensure we are not clobbering a dup container id request
+	if _, ok := lbf.ctrs[in.ContainerID]; ok {
```

Comment:
Should we check this in the beginning of this function instead?

Reply:
I think to move it up we would have to put most of the function under the mutex lock.  It is isolate at the bottom after the container object has been created so that we can check for duplication at the same time we do the assignment and to minimize the timing and code run under the lock.  If we move the dup check up, we would either have to check again anyway inside the lock (since concurrency is allowed) or lock before we create the container which could take time with creating volumes via the cache manager.  

I have tweaked the code a bit to tighten up the release and unlock handling, does that look better?